File: backend/app/schemas/leaderboard.py
"""
Leaderboard schemas.
Defines data models for leaderboard queries and responses.
Requirements: 5.5, 5.6, 5.8
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import Field

from app.schemas.base import BaseSchema


class LeaderboardCategory(str, Enum):
    """Available leaderboard categories."""
    WINS = "wins"
    WIN_RATE = "win_rate"
    TOTAL_SCORE = "total_score"
    KILLS = "kills"
    KD_RATIO = "kd_ratio"
    ACCURACY = "accuracy"
    FASTEST_THINKER = "fastest_thinker"
    ANSWER_RATE = "answer_rate"
    WIN_STREAK = "win_streak"
    ELO = "elo"  # New ELO-based category


class RankTier(str, Enum):
    """
    Rank tiers based on ELO rating.
    Requirements: 5.5
    """
    BRONZE = "Bronze"
    SILVER = "Silver"
    GOLD = "Gold"
    PLATINUM = "Platinum"
    DIAMOND = "Diamond"
    MASTER = "Master"
    GRANDMASTER = "Grandmaster"


# ============================================
# Legacy Leaderboard Schemas
# ============================================

class LeaderboardEntry(BaseSchema):
    """Single entry in a leaderboard."""
    rank: int
    user_id: str
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None
    stat_value: float
    secondary_stat: Optional[float] = None
    secondary_label: Optional[str] = None


class LeaderboardResponse(BaseSchema):
    """Leaderboard API response."""
    category: LeaderboardCategory
    entries: List[LeaderboardEntry]
    total_eligible: int
    page: int
    page_size: int
    minimum_requirement: Optional[str] = None


class UserRankResponse(BaseSchema):
    """User's rank in a specific category."""
    category: LeaderboardCategory
    rank: Optional[int] = None  # None if not eligible
    stat_value: float
    eligible: bool
    requirement_met: bool
    requirement: Optional[str] = None


class LeaderboardQueryParams(BaseSchema):
    """Query parameters for leaderboard requests."""
    limit: int = 10
    offset: int = 0


# ============================================
# ELO Rating Schemas (Requirements: 5.5, 5.6, 5.8)
# ============================================

class PlayerRating(BaseSchema):
    """
    Player's ELO rating.
    Requirements: 5.5, 5.6
    """
    user_id: str = Field(..., description="User UUID")
    display_name: Optional[str] = Field(None, description="Player display name")
    avatar_url: Optional[str] = Field(None, description="Player avatar URL")
    current_elo: int = Field(default=1200, description="Current ELO rating")
    peak_elo: int = Field(default=1200, description="Highest ELO achieved")
    current_tier: RankTier = Field(default=RankTier.GOLD, description="Current rank tier")
    tier_rank: Optional[int] = Field(None, description="Rank within tier")
    win_rate: float = Field(default=0.0, description="Win rate percentage")
    games_played: int = Field(default=0, description="Total ranked games played")
    last_match_date: Optional[datetime] = Field(None, description="Last match timestamp")


class ELOLeaderboardEntry(BaseSchema):
    """
    Single entry in ELO leaderboard.
    Requirements: 5.6
    """
    rank: int = Field(..., description="Global rank position")
    user_id: str = Field(..., description="User UUID")
    display_name: Optional[str] = Field(None, description="Player display name")
    avatar_url: Optional[str] = Field(None, description="Player avatar URL")
    elo: int = Field(..., description="Current ELO rating")
    tier: RankTier = Field(..., description="Rank tier")
    win_rate: float = Field(default=0.0, description="Win rate percentage")
    games_played: int = Field(default=0, description="Total games played")


class ELOLeaderboardResponse(BaseSchema):
    """
    ELO leaderboard API response.
    Requirements: 5.6
    """
    entries: List[ELOLeaderboardEntry] = Field(default_factory=list)
    total_players: int = Field(default=0, description="Total ranked players")
    page: int = Field(default=1)
    page_size: int = Field(default=100)
    region: Optional[str] = Field(None, description="Region filter if applied")


class UserELORankResponse(BaseSchema):
    """
    User's ELO rank with nearby players.
    Requirements: 5.8
    """
    rank: int = Field(..., description="User's global rank")
    total_players: int = Field(..., description="Total ranked players")
    rating: PlayerRating = Field(..., description="User's full rating info")
    nearby_players: List[ELOLeaderboardEntry] = Field(
        default_factory=list, 
        description="Players ±5 positions"
    )


# ============================================
# Match Result Schemas (for ELO calculation)
# ============================================

class MatchResult(BaseSchema):
    """
    Match result for ELO calculation.
    Requirements: 5.3
    """
    match_id: str = Field(..., description="Match UUID")
    player1_id: str = Field(..., description="Player 1 UUID")
    player2_id: str = Field(..., description="Player 2 UUID")
    winner_id: Optional[str] = Field(None, description="Winner UUID (None for draw)")
    duration_seconds: Optional[int] = Field(None, description="Match duration")
    player1_pre_elo: int = Field(..., description="Player 1 ELO before match")
    player2_pre_elo: int = Field(..., description="Player 2 ELO before match")
    player1_post_elo: int = Field(..., description="Player 1 ELO after match")
    player2_post_elo: int = Field(..., description="Player 2 ELO after match")
    elo_delta_p1: int = Field(..., description="ELO change for player 1")
    elo_delta_p2: int = Field(..., description="ELO change for player 2")
    played_at: datetime = Field(default_factory=datetime.utcnow)


class ELOCalculationResult(BaseSchema):
    """Result of ELO calculation for a match."""
    player1_new_elo: int
    player2_new_elo: int
    player1_delta: int
    player2_delta: int
    player1_new_tier: RankTier
    player2_new_tier: RankTier


# ============================================
# Seasonal Leaderboard Schemas
# ============================================

class SeasonalLeaderboardEntry(BaseSchema):
    """Entry in a seasonal leaderboard."""
    rank: int
    user_id: str
    display_name: Optional[str] = None
    avatar_url: Optional[str] = None
    season_elo: int
    peak_season_elo: int
    tier: RankTier
    games_played: int
    win_rate: float


class SeasonalLeaderboardResponse(BaseSchema):
    """Seasonal leaderboard response."""
    season_id: str
    season_number: int
    season_name: str
    entries: List[SeasonalLeaderboardEntry] = Field(default_factory=list)
    total_players: int = 0
    is_active: bool = True

File: frontend/src/types/leaderboard.ts
// Leaderboard types
// Requirements: 5.5, 5.6

export type LeaderboardCategory =
  | 'wins'
  | 'win_rate'
  | 'total_score'
  | 'kills'
  | 'kd_ratio'
  | 'accuracy'
  | 'fastest_thinker'
  | 'answer_rate'
  | 'win_streak'
  | 'elo'

/**
 * Rank tiers based on ELO rating.
 * Requirements: 5.5
 * Note: Backend uses PascalCase ('Bronze'), frontend uses lowercase ('bronze')
 */
export type RankTier = 'bronze' | 'silver' | 'gold' | 'platinum' | 'diamond' | 'master' | 'grandmaster';

// Backend tier names (PascalCase)
export type BackendRankTier = 'Bronze' | 'Silver' | 'Gold' | 'Platinum' | 'Diamond' | 'Master' | 'Grandmaster';

export const RANK_TIERS: Record<RankTier, { min: number; max: number; icon: string; color: string }> = {
  bronze: { min: 100, max: 799, icon: '🥉', color: '#cd7f32' },
  silver: { min: 800, max: 1199, icon: '🥈', color: '#c0c0c0' },
  gold: { min: 1200, max: 1599, icon: '🥇', color: '#ffd700' },
  platinum: { min: 1600, max: 1999, icon: '💎', color: '#60A5FA' },
  diamond: { min: 2000, max: 2399, icon: '💠', color: '#b9f2ff' },
  master: { min: 2400, max: 2799, icon: '👑', color: '#ff6b6b' },
  grandmaster: { min: 2800, max: 3000, icon: '🏆', color: '#ffd700' },
};

/**
 * Normalize backend tier name to frontend format.
 * Backend returns PascalCase ('Bronze'), frontend uses lowercase ('bronze')
 */
export function normalizeRankTier(tier: string | BackendRankTier): RankTier {
  return tier.toLowerCase() as RankTier;
}

/**
 * Get rank tier from ELO rating.
 */
export function getRankTier(elo: number): RankTier {
  if (elo >= 2800) return 'grandmaster';
  if (elo >= 2400) return 'master';
  if (elo >= 2000) return 'diamond';
  if (elo >= 1600) return 'platinum';
  if (elo >= 1200) return 'gold';
  if (elo >= 800) return 'silver';
  return 'bronze';
}

/**
 * Get rank tier info with safe normalization from backend response.
 */
export function getRankTierInfo(tierOrElo: string | number): { tier: RankTier; info: typeof RANK_TIERS[RankTier] } {
  const tier = typeof tierOrElo === 'number' 
    ? getRankTier(tierOrElo) 
    : normalizeRankTier(tierOrElo);
  return { tier, info: RANK_TIERS[tier] };
}

/**
 * Player rating with ELO and tier.
 */
export interface PlayerRating {
  user_id: string;
  elo: number;
  tier: RankTier;
  peak_elo: number;
  wins: number;
  losses: number;
}

export interface LeaderboardEntry {
  rank: number
  user_id: string
  display_name: string | null
  avatar_url: string | null
  stat_value: number
  secondary_stat: number | null
  secondary_label: string | null
}

export interface LeaderboardResponse {
  category: LeaderboardCategory
  entries: LeaderboardEntry[]
  total_eligible: number
  page: number
  page_size: number
  minimum_requirement: string | null
}

export interface UserRankResponse {
  category: LeaderboardCategory
  rank: number | null
  stat_value: number
  eligible: boolean
  requirement_met: boolean
  requirement: string | null
}

// ELO-specific types
export interface ELOLeaderboardEntry {
  rank: number
  user_id: string
  display_name: string | null
  avatar_url: string | null
  elo: number
  tier: BackendRankTier
  win_rate: number
  games_played: number
}

export interface ELOLeaderboardResponse {
  entries: ELOLeaderboardEntry[]
  total_players: number
  page: number
  page_size: number
  region?: string
}

export interface ELOPlayerRating {
  user_id: string
  display_name: string | null
  avatar_url: string | null
  current_elo: number
  peak_elo: number
  current_tier: BackendRankTier
  win_rate: number
  last_match_date: string | null
}

export interface UserELORankResponse {
  rank: number
  total_players: number
  rating: ELOPlayerRating
  nearby_players: ELOLeaderboardEntry[]
}

// Category metadata for display
export interface CategoryMeta {
  id: LeaderboardCategory
  name: string
  icon: string
  description: string
  format: (value: number) => string
  secondaryFormat?: (value: number) => string
  gradient: string
}

export const CATEGORY_META: Record<LeaderboardCategory, CategoryMeta> = {
  wins: {
    id: 'wins',
    name: 'Most Wins',
    icon: '🏆',
    description: 'Total games won',
    format: (v) => v.toLocaleString(),
    secondaryFormat: (v) => `${v} games`,
    gradient: 'from-amber-500 to-orange-600',
  },
  win_rate: {
    id: 'win_rate',
    name: 'Win Rate',
    icon: '📈',
    description: 'Highest win percentage (10+ games)',
    format: (v) => `${v.toFixed(1)}%`,
    secondaryFormat: (v) => `${v} games`,
    gradient: 'from-emerald-500 to-emerald-600',
  },
  total_score: {
    id: 'total_score',
    name: 'Total Score',
    icon: '⭐',
    description: 'Cumulative points earned',
    format: (v) => v.toLocaleString(),
    secondaryFormat: (v) => `${v} games`,
    gradient: 'from-indigo-500 to-indigo-600',
  },
  kills: {
    id: 'kills',
    name: 'Most Kills',
    icon: '💀',
    description: 'Total eliminations',
    format: (v) => v.toLocaleString(),
    secondaryFormat: (v) => `${v} deaths`,
    gradient: 'from-red-500 to-rose-600',
  },
  kd_ratio: {
    id: 'kd_ratio',
    name: 'K/D Ratio',
    icon: '⚔️',
    description: 'Kill/Death ratio (10+ deaths)',
    format: (v) => v.toFixed(2),
    secondaryFormat: (v) => `${v} kills`,
    gradient: 'from-orange-500 to-red-600',
  },
  accuracy: {
    id: 'accuracy',
    name: 'Accuracy',
    icon: '🎯',
    description: 'Shot accuracy (100+ shots)',
    format: (v) => `${v.toFixed(1)}%`,
    secondaryFormat: (v) => `${v} shots`,
    gradient: 'from-blue-500 to-indigo-600',
  },
  fastest_thinker: {
    id: 'fastest_thinker',
    name: 'Fastest Thinker',
    icon: '⚡',
    description: 'Quickest avg answer time (50+ correct)',
    format: (v) => `${(v / 1000).toFixed(2)}s`,
    secondaryFormat: (v) => `${v} correct`,
    gradient: 'from-yellow-500 to-amber-600',
  },
  answer_rate: {
    id: 'answer_rate',
    name: 'Answer Rate',
    icon: '🧠',
    description: 'Correct answer percentage (100+ questions)',
    format: (v) => `${v.toFixed(1)}%`,
    secondaryFormat: (v) => `${v} questions`,
    gradient: 'from-pink-500 to-rose-600',
  },
  win_streak: {
    id: 'win_streak',
    name: 'Win Streak',
    icon: '🔥',
    description: 'Best consecutive wins',
    format: (v) => v.toLocaleString(),
    secondaryFormat: (v) => `${v} current`,
    gradient: 'from-orange-500 to-yellow-500',
  },
  elo: {
    id: 'elo',
    name: 'ELO Rating',
    icon: '🏅',
    description: 'Competitive ranking',
    format: (v) => v.toLocaleString(),
    secondaryFormat: (v) => `Peak: ${v}`,
    gradient: 'from-indigo-500 to-indigo-600',
  },
}

export const ALL_CATEGORIES: LeaderboardCategory[] = [
  'elo',
  'wins',
  'win_rate',
  'total_score',
  'kills',
  'kd_ratio',
  'accuracy',
  'fastest_thinker',
  'answer_rate',
  'win_streak',
]
